### ISSUE TITLE
The compilation of the golang to mips failed

### ISSUE THREAD

**Issue by welkin22** [herumi/bls-eth-go-binary#62]
I am trying to compile a golang program in the docker image golang:1.21.3-alpine3.18:
```
RUN --mount=type=cache,target=/root/.cache/go-build cd op-program && make op-program-client-mips  \
    env CC=mips-linux-musl-gcc CGO_ENABLED=1 GOOS=linux GOARCH=mips GOMIPS=softfloat GITCOMMIT=$GIT_COMMIT GITDATE=$GIT_DATE VERSION="$OP_PROGRAM_VERSION"
```
The command for "make" is:
```
op-program-client-mips:
	env GO111MODULE=on GOOS=linux GOARCH=mips GOMIPS=softfloat go build -v $(LDFLAGS) -o ./bin/op-program-client.elf ./client/cmd/main.go
	# verify output with: readelf -h bin/op-program-client.elf
	# result is mips32, big endian, R3000
```
The Golang program depends on github.com/herumi/bls-eth-go-binary v0.0.0-20210917013441-d37c07cfda4e
I got the following error:
```
 > [builder 15/18] RUN --mount=type=cache,target=/root/.cache/go-build cd op-program && make op-program-client-mips      env CC=mips-linux-musl-gcc CGO_ENABLED=1 GOOS=linux GOARCH=mips GOMIPS=softfloat GITCOMMIT=$GIT_COMMIT GITDATE=$GIT_DATE VERSION="v0.0.0":
0.109 env GO111MODULE=on GOOS=linux GOARCH=mips GOMIPS=softfloat go build -v -ldflags "-X main.GitCommit= -X main.GitDate= -X github.com/ethereum-optimism/optimism/op-program/version.Version=v0.0.0 -X github.com/ethereum-optimism/optimism/op-program/version.Meta=" -o ./bin/op-program-client.elf ./client/cmd/main.go
2.538 # command-line-arguments
2.538 /usr/local/go/pkg/tool/linux_arm64/link: running mips-linux-musl-gcc failed: exit status 1
2.538 /app/mips-linux-musl-cross/bin/../lib/gcc/mips-linux-musl/11.2.1/../../../../mips-linux-musl/bin/ld: cannot find -lbls384_256
2.538 collect2: error: ld returned 1 exit status
2.538 
2.540 make: *** [Makefile:24: op-program-client-mips] Error 1
```
I guess it's because there is no corresponding libbls384_256.a file in the bls/lib folder.
I tried to generate this file in a docker container based on the instructions in the readme because my computer is a Mac and cannot generate the Linux version of this file.
```
RUN git clone --recursive https://github.com/herumi/bls-go-binary
RUN cd bls-go-binary && make
RUN cd bls-go-binary && env CXX=clang++ BIT=32 ARCH=mips _OS=linux _ARCH=mips make MCL_USE_GMP=0 UNIT=4 CFLAGS_USER="-target mips-linux -fPIC"
```
I export the newly generated file and put it in my own fork repository (https://github.com/welkin22/bls-eth-go-binary/tree/opbnb_fix). I modified link.go and added the following line:
```
#cgo linux,mips LDFLAGS:-L${SRCDIR}/lib/linux/mips
```
I got the following error using my own repository code:
```
#19 [builder 15/18] RUN --mount=type=cache,target=/root/.cache/go-build cd op-program && make op-program-client-mips      env CC=mips-linux-musl-gcc CGO_ENABLED=1 GOOS=linux GOARCH=mips GOMIPS=softfloat GITCOMMIT=$GIT_COMMIT GITDATE=$GIT_DATE VERSION="v0.0.0"
#19 0.087 env GO111MODULE=on GOOS=linux GOARCH=mips GOMIPS=softfloat go build -v -ldflags "-X main.GitCommit= -X main.GitDate= -X github.com/ethereum-optimism/optimism/op-program/version.Version=v0.0.0 -X github.com/ethereum-optimism/optimism/op-program/version.Meta=" -o ./bin/op-program-client.elf ./client/cmd/main.go
#19 0.545 github.com/herumi/bls-eth-go-binary/bls
#19 12.97 github.com/prysmaticlabs/prysm/v4/crypto/bls/herumi
#19 12.98 github.com/prysmaticlabs/prysm/v4/crypto/bls
#19 13.03 command-line-arguments
#19 15.01 # command-line-arguments
#19 15.01 /usr/local/go/pkg/tool/linux_arm64/link: running mips-linux-musl-gcc failed: exit status 1
#19 15.01 /app/mips-linux-musl-cross/bin/../lib/gcc/mips-linux-musl/11.2.1/../../../../mips-linux-musl/bin/ld: skipping incompatible /go/pkg/mod/github.com/welkin22/bls-eth-go-binary@v0.0.2-opbnb/bls/lib/linux/mips/libbls384_256.a when searching for -lbls384_256
#19 15.01 /app/mips-linux-musl-cross/bin/../lib/gcc/mips-linux-musl/11.2.1/../../../../mips-linux-musl/bin/ld: cannot find -lbls384_256
#19 15.01 collect2: error: ld returned 1 exit status
#19 15.01 
#19 15.02 make: *** [Makefile:24: op-program-client-mips] Error 1
#19 ERROR: process "/bin/sh -c cd op-program && make op-program-client-mips      env CC=mips-linux-musl-gcc CGO_ENABLED=1 GOOS=linux GOARCH=mips GOMIPS=softfloat GITCOMMIT=$GIT_COMMIT GITDATE=$GIT_DATE VERSION=\"$OP_PROGRAM_VERSION\"" did not complete successfully: exit code: 2
```
It seems that the files I added were ignored due to compatibility issues.
Where did the mistake occur? Please help take a look, thank you very much

**Comment by herumi** [comment]
I am not familiar with cross-compilation in Go but I was able to cross-compile the lib on mips and run it in QEMU.
After installing `g++-13-mipsisa32r6el-linux-gnu` etc on Ubuntu, and in `bls-eth-go-binary/src/bls`, type

```
make -f Makefile.onelib ARCH=mipsisa32r6el CLANG_TARGET=mipsisa32r6el BIT=32 CXX=clang++-17 ETH_CFLAGS=-DBLS_ETH
mipsisa32r6el-linux-gnu-gcc-13 sample/minsample.c -I include/ -I mcl/include/ bls/lib/linux/libbls384_256.a -DBLS_ETH
env QEMU_LD_PREFIX=/usr/mipsisa32r6el-linux-gnu/ ./a.out
```
Then it shows

```
BLS_ETH mode
verify correct message 1
verify wrong message 0
recover test1 is ok
recover test2 is ok
```

`mipsisa32r6el` should be changed according to your environment.

**Comment by welkin22** [comment]
@herumi hi, I tried the command you sent: `make -f Makefile.onelib ARCH=mipsisa32r6el CLANG_TARGET=mipsisa32r6el BIT=32 CXX=clang++-17 ETH_CFLAGS=-DBLS_ETH `
This is my version:
```
RUN git clone --recursive https://github.com/herumi/bls-eth-go-binary.git
RUN cd bls-eth-go-binary/src/bls && make -f Makefile.onelib ARCH=mips CLANG_TARGET=mips BIT=32 CXX=clang++-17 ETH_CFLAGS=-DBLS_ETH
```
I get the following error:
```
#18 [builder 14/15] RUN cd bls-eth-go-binary/src/bls && make -f Makefile.onelib ARCH=mips CLANG_TARGET=mips BIT=32 CXX=clang++-17 ETH_CFLAGS=-DBLS_ETH
#18 0.107 clang++-17 -c src/bls_c384_256.cpp -o obj/bls_c384_256.o -O3 -fno-exceptions -fno-threadsafe-statics -fno-rtti -fno-stack-protector -DNDEBUG -DMCL_DONT_USE_OPENSSL -DMCL_SIZEOF_UNIT=4 -DMCL_MAX_BIT_SIZE=384 -DCYBOZU_DONT_USE_EXCEPTION -DCYBOZU_DONT_USE_STRING -D_FORTIFY_SOURCE=0 -I.//include -I.//mcl/include -DBLS_ETH  -std=c++03 -fpic -target mips -DMCL_USE_LLVM=1
#18 0.203 In file included from src/bls_c384_256.cpp:3:
#18 0.203 In file included from src/bls_c_impl.hpp:6:
#18 0.203 In file included from .//include/bls/bls.h:10:
#18 0.203 In file included from .//mcl/include/mcl/bn.h:62:
#18 0.203 In file included from .//mcl/include/mcl/curve_type.h:61:
#18 0.203 In file included from .//mcl/include/cybozu/inttype.hpp:33:
#18 0.203 In file included from /usr/include/unistd.h:37:
#18 0.203 /usr/include/bits/alltypes.h:58:24: error: typedef redefinition with different types ('unsigned long' vs 'unsigned int')
#18 0.203    58 | typedef unsigned _Addr size_t;
#18 0.203       |                        ^
#18 0.203 /usr/lib/llvm17/lib/clang/17/include/stddef.h:46:23: note: previous definition is here
#18 0.203    46 | typedef __SIZE_TYPE__ size_t;
#18 0.203       |                       ^
```
It seems there is a definition conflict, the docker image I am using is defined as follows:
```
FROM golang:1.21.5-alpine3.19 as builder

RUN apk add --no-cache make gcc linux-headers git jq bash wget clang
```
Do you know how to solve this problem? I tried searching the internet but didn't get a useful answer.

**Comment by herumi** [comment]
The error seems to be a compiler version issue.
I installed `clang++-17` and specified it. But you did not install it. How about trying `CXX=clang++`?

**Comment by welkin22** [comment]
> CXX=clang++

same result:
```
#18 [builder 14/15] RUN cd bls-eth-go-binary/src/bls && make -f Makefile.onelib ARCH=mips CLANG_TARGET=mips BIT=32 CXX=clang++ ETH_CFLAGS=-DBLS_ETH
#18 0.099 clang++ -c src/bls_c384_256.cpp -o obj/bls_c384_256.o -O3 -fno-exceptions -fno-threadsafe-statics -fno-rtti -fno-stack-protector -DNDEBUG -DMCL_DONT_USE_OPENSSL -DMCL_SIZEOF_UNIT=4 -DMCL_MAX_BIT_SIZE=384 -DCYBOZU_DONT_USE_EXCEPTION -DCYBOZU_DONT_USE_STRING -D_FORTIFY_SOURCE=0 -I.//include -I.//mcl/include -DBLS_ETH  -std=c++03 -fpic -target mips -DMCL_USE_LLVM=1
#18 0.259 In file included from src/bls_c384_256.cpp:3:
#18 0.259 In file included from src/bls_c_impl.hpp:6:
#18 0.259 In file included from .//include/bls/bls.h:10:
#18 0.259 In file included from .//mcl/include/mcl/bn.h:62:
#18 0.259 In file included from .//mcl/include/mcl/curve_type.h:61:
#18 0.259 In file included from .//mcl/include/cybozu/inttype.hpp:33:
#18 0.259 In file included from /usr/include/unistd.h:37:
#18 0.259 /usr/include/bits/alltypes.h:58:24: error: typedef redefinition with different types ('unsigned long' vs 'unsigned int')
#18 0.259    58 | typedef unsigned _Addr size_t;
#18 0.259       |                        ^
#18 0.259 /usr/lib/llvm17/lib/clang/17/include/stddef.h:46:23: note: previous definition is here
#18 0.259    46 | typedef __SIZE_TYPE__ size_t;
#18 0.259       |                       ^
```

And I has confirmed the `clang++-17` is already be installed:
```
#6 [builder  2/15] RUN apk add --no-cache make gcc musl-dev linux-headers git jq bash wget clang libstdc++-dev
#6 0.173 fetch https://dl-cdn.alpinelinux.org/alpine/v3.19/main/aarch64/APKINDEX.tar.gz
#6 2.068 fetch https://dl-cdn.alpinelinux.org/alpine/v3.19/community/aarch64/APKINDEX.tar.gz
#6 3.122 (1/41) Installing ncurses-terminfo-base (6.4_p20231125-r0)
#6 3.182 (2/41) Installing libncursesw (6.4_p20231125-r0)
#6 3.261 (3/41) Installing readline (8.2.1-r2)
#6 3.336 (4/41) Installing bash (5.2.21-r0)
#6 3.613 Executing bash-5.2.21-r0.post-install
#6 3.617 (5/41) Installing clang17-headers (17.0.5-r0)
#6 4.140 (6/41) Installing libffi (3.4.4-r3)
#6 4.202 (7/41) Installing libgcc (13.2.1_git20231014-r0)
#6 4.265 (8/41) Installing libstdc++ (13.2.1_git20231014-r0)
#6 4.568 (9/41) Installing xz-libs (5.4.5-r0)
#6 4.652 (10/41) Installing libxml2 (2.11.8-r0)
#6 4.817 (11/41) Installing zstd-libs (1.5.5-r8)
#6 4.980 (12/41) Installing llvm17-libs (17.0.5-r0)
#6 15.38 (13/41) Installing clang17-libs (17.0.5-r0)
#6 22.95 (14/41) Installing fortify-headers (1.1-r3)
#6 23.00 (15/41) Installing jansson (2.14-r4)
#6 23.06 (16/41) Installing binutils (2.41-r0)
#6 23.86 (17/41) Installing libgomp (13.2.1_git20231014-r0)
#6 23.95 (18/41) Installing libatomic (13.2.1_git20231014-r0)
#6 24.00 (19/41) Installing gmp (6.3.0-r0)
#6 24.09 (20/41) Installing isl26 (0.26-r1)
#6 24.50 (21/41) Installing mpfr4 (4.2.1-r0)
#6 24.67 (22/41) Installing mpc1 (1.3.1-r1)
#6 24.73 (23/41) Installing gcc (13.2.1_git20231014-r0)
#6 33.92 (24/41) Installing libstdc++-dev (13.2.1_git20231014-r0)
#6 34.61 (25/41) Installing llvm17-linker-tools (17.0.5-r0)
#6 34.67 (26/41) Installing musl-dev (1.2.4_git20230717-r4)
#6 35.72 (27/41) Installing clang17 (17.0.5-r0)
#6 35.89 (28/41) Installing brotli-libs (1.1.0-r1)
#6 36.06 (29/41) Installing c-ares (1.27.0-r0)
#6 36.13 (30/41) Installing libunistring (1.1-r2)
#6 36.32 (31/41) Installing libidn2 (2.3.4-r4)
#6 36.39 (32/41) Installing nghttp2-libs (1.58.0-r0)
#6 36.47 (33/41) Installing libcurl (8.5.0-r0)
#6 36.59 (34/41) Installing libexpat (2.6.2-r0)
#6 36.69 (35/41) Installing pcre2 (10.42-r2)
#6 36.88 (36/41) Installing git (2.43.4-r0)
#6 38.11 (37/41) Installing oniguruma (6.9.9-r0)
#6 38.24 (38/41) Installing jq (1.7.1-r0)
#6 38.41 (39/41) Installing linux-headers (6.5-r0)
#6 39.00 (40/41) Installing make (4.4.1-r2)
#6 39.09 (41/41) Installing wget (1.21.4-r0)
```

**Comment by herumi** [comment]
Can you compile the following code with -target mips?

```c
#include <unistd.h>
#include <stdio.h>

int main() {
  puts("hello");
}
```
How about appending `CFLAGS_USER=-DMCL_STANDALNONE` option?

**Comment by welkin22** [comment]
> Can you compile the following code with -target mips?
> 
> ```c
> #include <unistd.h>
> #include <stdio.h>
> 
> int main() {
>   puts("hello");
> }
> ```
> 
> How about appending `CFLAGS_USER=-DMCL_STANDALNONE` option?

hi @herumi , after appending `CFLAGS_USER=-DMCL_STANDALNONE` option, the result is same.
I thought of a good way to facilitate your troubleshooting. Since I use docker, I easily created a repository: https://github.com/welkin22/for-bls-issue-try. After you clone this repository, you can reproduce my attempt to compile the libbls384_256.a file by using the command `docker build --output ./ -f Dockerfile-buildLib .`. My ultimate goal is to use the command `docker build --output ./ -f Dockerfile .` to get the main.elf file. The main.go file only contains the code from the bls-eth-go-binary repository. You can see the problems I am currently facing.

**Comment by welkin22** [comment]
@herumi hi, I just noticed that there is a spelling error in the `CFLAGS_USER=-DMCL_STANDALNONE` you sent me. It should be `CFLAGS_USER=-DMCL_STANDALONE`. After the modification, the error that occurred previously no longer occurs, but it has moved to another place:
```
#18 0.126 In file included from src/bls_c384_256.cpp:3:
#18 0.126 In file included from src/bls_c_impl.hpp:8:
#18 0.126 In file included from .//mcl/include/mcl/impl/bn_c_impl.hpp:13:
#18 0.126 In file included from .//mcl/include/mcl/bls12_381.hpp:11:
#18 0.126 In file included from .//mcl/include/mcl/bn.hpp:11:
#18 0.126 In file included from .//mcl/include/mcl/fp_tower.hpp:9:
#18 0.126 In file included from .//mcl/include/mcl/fp.hpp:29:
#18 0.126 In file included from .//mcl/include/cybozu/stream.hpp:15:
#18 0.126 In file included from /usr/include/memory.h:1:
#18 0.126 In file included from /usr/include/string.h:25:
#18 0.126 /usr/include/bits/alltypes.h:58:24: error: typedef redefinition with different types ('unsigned long' vs 'unsigned int')
#18 0.126    58 | typedef unsigned _Addr size_t;
#18 0.126       |                        ^
#18 0.126 /usr/lib/llvm17/lib/clang/17/include/stddef.h:46:23: note: previous definition is here
#18 0.126    46 | typedef __SIZE_TYPE__ size_t;
#18 0.126       |                       ^
```

**Comment by herumi** [comment]
Oh, sorry, but did you try to compile this code?
```c
#include <unistd.h>
#include <stdio.h>

int main() {
  puts("hello");
}
```
If it fails, then the problem is not `bls-eth-go-binary`.

**Comment by welkin22** [comment]
> Oh, sorry, but did you try to compile this code?
> 
> ```c
> #include <unistd.h>
> #include <stdio.h>
> 
> int main() {
>   puts("hello");
> }
> ```
> 
> If it fails, then the problem is not `bls-eth-go-binary`.

@herumi hi, I have solved this problem. The reason for the problem is that the docker environment I used has the musl-dev library, which introduces a series of header files in /usr/include that conflict with the header files installed by clang. I just need to add `-isystem /usr/include` to CFLAGS_USER, specifying that clang++ treats the header files in /usr/include as system header files with higher priority to resolve this conflict.
After solving this conflict issue, I found that when compiling golang code, the compiler complained about the undefined `blsSecretKeySetByCSPRNG` and `blsSetRandFunc`. It turned out that configuring `DMCL_STANDALONE` would enable `MCL_DONT_USE_CSPRNG`, causing these two methods to be ignored. Therefore, I cannot configure `DMCL_STANDALONE`.
After removing `DMCL_STANDALONE` and adding `-isystem /usr/include`, I succeeded, and the compiled `libbls384_256.a` file works fine. Thank you for taking the time to help troubleshoot the issue.
